src/function/table/arrow.cpp

```diff
@@ -0,0 +1,260 @@
+		}
+		if (array.dictionary) {
+			throw NotImplementedException("arrow_scan: dictionary vectors not supported yet");
+		}
+		if (array.null_count != 0 && array.buffers[0]) {
+			auto &nullmask = FlatVector::Nullmask(output.data[col_idx]);
+
+			auto bit_offset = data.chunk_offset + array.offset;
+			auto n_bitmask_bytes = (output.size() + 8 - 1) / 8;
+
+			if (bit_offset % 8 == 0) {
+				// just memcpy nullmask
+				memcpy(&nullmask, (uint8_t *)array.buffers[0] + bit_offset / 8, n_bitmask_bytes);
+			} else {
+				// need to re-align nullmask :/
+				bitset<STANDARD_VECTOR_SIZE + 8> temp_nullmask;
+				memcpy(&temp_nullmask, (uint8_t *)array.buffers[0] + bit_offset / 8, n_bitmask_bytes + 1);
+
+				temp_nullmask >>= (bit_offset % 8); // why this has to be a right shift is a mystery to me
+				memcpy(&nullmask, (data_ptr_t)&temp_nullmask, n_bitmask_bytes);
+			}
+			nullmask.flip(); // arrow uses inverse nullmask logic
+		}
+
+		switch (output.data[col_idx].type.id()) {
+		case LogicalTypeId::BOOLEAN:
+		case LogicalTypeId::TINYINT:
+		case LogicalTypeId::SMALLINT:
+		case LogicalTypeId::INTEGER:
+		case LogicalTypeId::FLOAT:
+		case LogicalTypeId::DOUBLE:
+		case LogicalTypeId::BIGINT:
+		case LogicalTypeId::HUGEINT:
+			FlatVector::SetData(output.data[col_idx],
+			                    (data_ptr_t)array.buffers[1] + GetTypeIdSize(output.data[col_idx].type.InternalType()) *
+			                                                       (data.chunk_offset + array.offset));
+			break;
+
+		case LogicalTypeId::VARCHAR: {
+			auto offsets = (uint32_t *)array.buffers[1] + array.offset + data.chunk_offset;
+			auto cdata = (char *)array.buffers[2];
+
+			for (idx_t row_idx = 0; row_idx < output.size(); row_idx++) {
+				if (FlatVector::Nullmask(output.data[col_idx])[row_idx]) {
+					continue;
+				}
+				auto cptr = cdata + offsets[row_idx];
+				auto str_len = offsets[row_idx + 1] - offsets[row_idx];
+
+				auto utf_type = Utf8Proc::Analyze(cptr, str_len);
+				switch (utf_type) {
+				case UnicodeType::ASCII:
+					FlatVector::GetData<string_t>(output.data[col_idx])[row_idx] =
+					    StringVector::AddString(output.data[col_idx], cptr, str_len);
+					break;
+				case UnicodeType::UNICODE:
+					// this regrettably copies to normalize
+					FlatVector::GetData<string_t>(output.data[col_idx])[row_idx] =
+					    StringVector::AddString(output.data[col_idx], Utf8Proc::Normalize(string(cptr, str_len)));
+
+					break;
+				case UnicodeType::INVALID:
+					throw runtime_error("Invalid UTF8 string encoding");
+				}
+			}
+
+			break;
+		} // TODO timestamps in duckdb are subject to change
+		case LogicalTypeId::TIMESTAMP: {
+			auto src_ptr = (uint64_t *)array.buffers[1] + data.chunk_offset;
+			auto tgt_ptr = (timestamp_t *)FlatVector::GetData(output.data[col_idx]);
+
+			for (idx_t row = 0; row < output.size(); row++) {
+				auto source_idx = data.chunk_offset + row;
+
+				auto ms = src_ptr[source_idx] / 1000000; // nanoseconds
```

Comment:
Need to use the time unit from the schema -- this will be incorrect for units other than nanoseconds?

Reply:
yes we need to do another iteration on the many time types that arrow seems to have